### DISCUSSION TITLE
Can playback simulate recieved frames? For testing scripts

### DISCUSSION THREAD

**Discussion by davearko** [collin80/SavvyCAN#557] [Ideas]
I'm not finding an easy way to test custom java-scripts that are listening for messages and sending responses.

I would like to playback the (filtered) log from a recording and have those messages act as if they were being received again (not sent out on the bus).  This would allow me to test my script and get the bugs worked out before connecting to the car.

I'll probably need to buy a 2nd CanDue to do this kind of off-line testing.   Please let me know if there is a better way.

I love SavvyCAN and all the work you put into it!  Thanks!
--Dave

**Comment by collin80** [comment]
What you'll want to do is use virtual buses. All three of the operating systems should have a virtual driver. Go to the connection window, add a bus, use QT SerialBus, select the subtype "virtualcan", leave it at can0 and add it. 

Now, if you add just one you can send all you want to it but nothing comes back. The key is to add a second one. But, use the exact same can0. Don't let it be can1. If two virtualcan sockets are attached to the same can number they'll act properly. When one sends the other will register receiving.

But, you will need to grab the continuous build I am just now having compiled. There was a crash bug in the code that gets potentially triggered when you have two virtual buses set up this way and you try to run a script. I'm digging into it further but it seems good enough for now for you to be able to use it.
